aspects/transaction.go

```diff
@@ -0,0 +1,125 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package aspects
+
+import (
+	"sync"
+)
+
+type DatabagRead func() (JSONDataBag, error)
+type DatabagWrite func(JSONDataBag) error
+
+// Transaction performs read and writes to a databag in an atomic way.
+type Transaction struct {
+	pristine JSONDataBag
+	deltas   []map[string]interface{}
+	schema   Schema
+
+	readDatabag  DatabagRead
+	writeDatabag DatabagWrite
+	mu           sync.RWMutex
+}
+
+// NewTransaction takes a getter and setter to read and write the databag.
+func NewTransaction(readDatabag DatabagRead, writeDatabag DatabagWrite, schema Schema) (*Transaction, error) {
```

Comment:
this is fine for now but later a transaction will have to deal with multiple aspect bundles/databags